### DISCUSSION TITLE
Many-to-many pivot table with same table using Orchid CRUD resource

### DISCUSSION THREAD

**Discussion by LemonPie22** [orchidsoftware/platform#2302] [Q&A]
I'm trying to implement a hierarchical relationship between products, using Orchid CRUD, where any product can have multiple sub-products (known as components) as well as multiple parent products.

For example, a car has an engine which has an oil cooler which has a bolt. The same bolt, which is a component of the oil cooler, could be used on other parts too.

To achieve this I have a main `products` table and a pivot table called `product_component`.

The `products` table has `id` and `name` columns.

The `product_component` table has `product_id` and `component_id` columns, both of which have foreign key relationships to the `products` table.

I'm struggling to show product components (the first step; I can do parents after) using an Orchid CRUD resource with a `Relation` field.

My `Product` model contains the following relationships:

```
public function products(): BelongsToMany
{
    return $this->belongsToMany(related: Product::class, table: 'product_component', foreignPivotKey: 'component_id', relatedPivotKey: 'product_id');
}

public function components(): BelongsToMany
{
    return $this->belongsToMany(related: Product::class, table: 'product_component', foreignPivotKey: 'product_id', relatedPivotKey: 'component_id');
}
```
And my `ProductResource` contains the following `FieldsRelation`:
```
use Orchid\Screen\Fields\Relation as FieldsRelation;

  FieldsRelation::make('product_id')
      ->fromModel(Product::class, 'name', 'id')
      ->multiple()
      ->title('Components'),
```
This allows me to add products in the UI:

![Screenshot 2022-07-09 at 16 05 10](https://user-images.githubusercontent.com/107488019/178111356-e555b9ef-82d8-4d8b-ab2b-3a43b4a00e44.png)

But as soon as I try to save I get this error:
```
SQLSTATE[42S22]: Column not found: 1054 Unknown column 'product_id' in 'field list'
update
  `products`
set
  `product_id` = [ "2" ],
  `products`.`updated_at` = 2022 -07 -09 15: 06: 12
where
  `id` = 1
```

I understand that I shouldn't be trying to save into the main Products table but can't work out how to save to the pivot table instead.

I've tried lots of things, and have read articles on here that touch on this topic, but can't seem to make any progress so I'd really appreciate some help.

**Comment by tabuna** [comment]
Something like this would look like saving the relationship:

```php
    /**
     * Action to create and update the model
     *
     * @param ResourceRequest $request
     * @param Model           $model
     */
    public function save(ResourceRequest $request, Model $model): void
    {
        $model->fill($request->all())->save();
        $model->products()->save($request->get('product_id'));
    }
```

**Reply by LemonPie22** [reply]
Thanks Alexandr. Does this go in the model or resource file?

**Reply by tabuna** [reply]
In the resource class, you can specify this and other behaviors

**Reply by LemonPie22** [reply]
I got this error:
`Method 'App\Orchid\Resources\ProductResource::save()' is not compatible with method 'Orchid\Crud\Resource::save()'`

**Reply by tabuna** [reply]
I think you have changed part of the code that I gave as an example. For example, you defined your model in expectations or specified the return type of the method. I don't have much information to be sure, but if so, you can use the `onSave` method, which shouldn't be such a problem.

**Reply by LemonPie22** [reply]
Thanks for you advice Alexandr. Much appreciated. 
I didn't change the code you posted at all.
I fixed the error by adding:
```
use Illuminate\Database\Eloquent\Model;
use Orchid\Crud\ResourceRequest;
```
Now it gets further but I'm getting this error on save:
```
Illuminate\Database\Eloquent\Relations\BelongsToMany::save(): Argument #1 ($model) must be of type Illuminate\Database\Eloquent\Model, array given, called in /var/www/html/app/Orchid/Resources/ProductResource.php on line 130
```
And line 130 is `$model->products()->save($request->get('product_id'));`
